src/lib/comms/sol-lwm2m-server.c

```diff
@@ -802,34 +812,113 @@ observer_entry_del_monitor(struct observer_entry *entry,
 {
     struct sol_lwm2m_server *server;
-    struct sol_network_link_addr servaddr = { .family = SOL_NETWORK_FAMILY_INET6,
-                                              .port = port };
+    struct sol_network_link_addr servaddr_coap = { .family = SOL_NETWORK_FAMILY_INET6,
+                                                   .port = coap_port };
+    struct sol_network_link_addr servaddr_dtls = { .family = SOL_NETWORK_FAMILY_INET6,
+                                                   .port = dtls_port };
     int r;
+    const char *sec_mode_str;
+    struct sol_lwm2m_security_psk *psk, *given_psk;
+    uint16_t i;
 
     SOL_LOG_INTERNAL_INIT_ONCE;
 
+    switch (sec_mode) {
+    case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
+        SOL_NULL_CHECK(known_psks, NULL);
+        sec_mode_str = "Pre-Shared Key";
+        break;
+    case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
+        sec_mode_str = "Raw Public Key";
+        SOL_WRN("Raw Public Key security mode is not supported yet.");
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
+        sec_mode_str = "Certificate";
+        SOL_WRN("Certificate security mode is not supported yet.");
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
+        sec_mode_str = "NoSec";
+        SOL_DBG("Using NoSec Security Mode (No DTLS).");
+        break;
+    default:
+        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
+        return NULL;
+    }
+
     server = calloc(1, sizeof(struct sol_lwm2m_server));
     SOL_NULL_CHECK(server, NULL);
 
-    server->coap = sol_coap_server_new(&servaddr, false);
+    server->coap = sol_coap_server_new(&servaddr_coap, false);
     SOL_NULL_CHECK_GOTO(server->coap, err_coap);
 
+    if (sec_mode != SOL_LWM2M_SECURITY_MODE_NO_SEC) {
+        if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+            sol_vector_init(&server->known_psks, sizeof(sol_lwm2m_security_psk));
+
+            SOL_VECTOR_FOREACH_IDX (known_psks, given_psk, i) {
+                psk = sol_vector_append(&server->known_psks);
+                SOL_NULL_CHECK_GOTO(psk, err_psk);
+                psk->id = sol_blob_ref(given_psk->id);
+                psk->key = sol_blob_ref(given_psk->key);
+            }
+        }
+
+        server->dtls_server = sol_coap_server_new(&servaddr_dtls, true);
+        SOL_NULL_CHECK_GOTO(server->dtls_server, err_dtls);
+
+        server->security = sol_lwm2m_server_security_add(server, sec_mode);
+        if (!server->security) {
+            if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+                SOL_VECTOR_FOREACH_IDX (&server->known_psks, psk, i) {
+                    sol_blob_unref(psk->id);
+                    sol_blob_unref(psk->key);
+                }
+                sol_vector_clear(&server->known_psks);
+            }
+            sol_coap_server_unref(server->dtls_server);
+            SOL_WRN("Could not enable %s security mode for LWM2M Server", sec_mode_str);
```

Comment:
Why should I? DTLS is not mandatory for LWM2M Servers... in this case, this one will just be a Server which listens on CoAP but not on DTLS. Clients who connect to it through insecure channels will be ok, clients who try to use secure channels will find a closed port in the network.